### ISSUE TITLE
[Bug] HiCache persistent storage can reload stale KV pages after online weight update

### ISSUE THREAD

**Issue by glaziermag** [sgl-project/sglang#26792]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When SGLang runs with HiCache persistent file storage enabled, `update_weights_from_tensor` with `flush_cache=True` clears the live GPU/host KV cache but does not invalidate the persistent HiCache file storage. A later request with the same prompt can reload KV pages that were computed under the previous model weights.

I confirmed this on current main commit `7ed53d15f357ea4d722c1980c2cb35e8367d8bb0` with `Qwen/Qwen3-30B-A3B` on a single NVIDIA A100 80GB.

Observed evidence:

- Request A used the original weights and wrote persistent HiCache pages.
- The online update succeeded with `flush_cache=True`, `abort_all_requests=True`, and `weight_version="step-1"`.
- The updated tensor was `model.layers.0.self_attn.qkv_proj.weight`, shape `[5120, 2048]`, set to CUDA bf16 zeros.
- HiCache file storage had 4166 files after A.
- HiCache file storage still had 4166 files after the weight update.
- 4166 old files survived the update; 0 new files were created by the update itself.
- Request B used the same prompt after the update and reported 4157 cached tokens from persistent storage:

```json
{
  "device": 0,
  "host": 0,
  "storage": 4157,
  "storage_backend": "HiCacheFile"
}
```

Controls after the same weight update did not use storage and agreed with each other:

- C: clear storage directory before the post-update request, cached tokens 0.
- D: disable HiCache storage backend before the post-update request, cached tokens 0.
- E: fresh server with updated weights and empty storage, cached tokens 0.

Checksums over output token IDs plus output logprobs/top-logprobs:

- B, storage intact after update: `aaaafc985422c90b9ce6f639d98fc55d19841b899cf9c5036481d1cd751aba70`
- C, storage cleared: `220c9966a442040d27d0a21802de2082fbe6973f775eb98c2178c35ef7db630b`
- D, storage disabled: `220c9966a442040d27d0a21802de2082fbe6973f775eb98c2178c35ef7db630b`
- E, fresh updated server: `220c9966a442040d27d0a21802de2082fbe6973f775eb98c2178c35ef7db630b`

The generated output token IDs happened to remain the same in this run, but logprobs/top-logprobs diverged. This still exposes a correctness issue: request B reused old-weight prefix KV and produced logits/logprobs different from all updated-weight controls.

Additional instrumentation showed that the HiCache file key suffix was model/parallelism based, not `weight_version` based:

```text
_qwen3-30b-a3b-hicache-debug_0_1
```

The same first storage keys appeared before and after the weight update. The read path reported storage files whose mtimes predated the update timestamp.

Static code audit from the same commit:

- `flush_cache_after_weight_update()` delegates to scheduler `flush_cache()` when `flush_cache=True`.
- Scheduler `flush_cache()` resets live radix/pool state and metrics, but does not clear the HiCache persistent file backend.
- HiCache file names include token/page hash plus model/TP/PP suffix, but no observed `weight_version` component.
- Persistent storage clearing exists as a separate admin path, not as part of weight-update flush.

Expected behavior:

After model weights change, KV cache computed under old weights must not be reused. `flush_cache=True` after an online weight update should invalidate every cache level that can affect generation under the new weights, including persistent HiCache storage. Reasonable fixes would be to clear persistent storage during weight-update flush, include `weight_version` or model identity in persistent storage keys, or reject/disable persistent HiCache storage across online weight updates until it is explicitly reset.

### Reproduction

I used a local validation harness equivalent to the following flow.

1. Clear the HiCache file storage directory.
2. Start SGLang with hierarchical cache and file storage:

```bash
export SGLANG_HICACHE_FILE_BACKEND_STORAGE_DIR=/home/gabe/hicache-test

python -m sglang.launch_server \
  --model-path Qwen/Qwen3-30B-A3B \
  --host 127.0.0.1 \
  --port 30000 \
  --dtype bfloat16 \
  --trust-remote-code \
  --served-model-name qwen3-30b-a3b-hicache-debug \
  --admin-api-key hicache-debug-admin \
  --enable-hierarchical-cache \
  --hicache-storage-backend file \
  --hicache-write-policy write_through \
  --hicache-storage-prefetch-policy wait_complete \
  --hicache-size 80 \
  --mem-fraction-static 0.88 \
  --context-length 8192 \
  --max-running-requests 1 \
  --disable-cuda-graph \
  --enable-deterministic-inference
```

3. Send request A with a deterministic long prompt, 4158 prompt tokens, `temperature=0`, and small `max_new_tokens`.
4. Wait until HiCache file storage is stable.
5. Apply an online weight update:

```python
from sglang.srt.utils import MultiprocessingSerializer
import requests
import torch

name = "model.layers.0.self_attn.qkv_proj.weight"
tensor = torch.zeros((5120, 2048), dtype=torch.bfloat16, device="cuda")
payload = {
    "serialized_named_tensors": [
        MultiprocessingSerializer.serialize([(name, tensor)], output_str=True)
    ],
    "load_format": "direct",
    "flush_cache": True,
    "abort_all_requests": True,
    "weight_version": "step-1",
    "torch_empty_cache": False,
}
requests.post(
    "http://127.0.0.1:30000/update_weights_from_tensor",
    json=payload,
    headers={"Authorization": "Bearer hicache-debug-admin"},
    timeout=600,
)
```

6. Send request B with the same prompt and sampling while keeping storage intact.
7. Compare with controls:
   - C: same update, but clear storage before the post-update request.
   - D: same update, but delete/disable the HiCache storage backend before the post-update request.
   - E: fresh updated server with empty storage.

The confirmation criteria were:

- B gets persistent storage hits for pages written before the update.
- Storage keys are identical before and after the update.
- B differs from the clear-storage/no-storage/fresh-updated controls.

All three criteria were met.

### Environment

The raw `python3 -m sglang.check_env` output was not retained before the GCP VM was deleted, but the run environment was:

- SGLang: editable checkout of `sgl-project/sglang` main at `7ed53d15f357ea4d722c1980c2cb35e8367d8bb0`
- Hardware: GCP `a2-ultragpu-1g`, single NVIDIA A100-SXM4-80GB
- Model: `Qwen/Qwen3-30B-A3B`
- Python: 3.10 paths in logs (`/home/gabe/.local/lib/python3.10/site-packages/...`)
- PyTorch/CUDA: CUDA-enabled PyTorch; server log included a warning that PyTorch was built with CUDA 13.0 while TorchAudio was built with CUDA 12.9
- Attention backend selected by server: prefill `fa3`, decode `flashinfer`
- Dtype: `bfloat16`
- Context length: 8192
- HiCache settings: hierarchical cache enabled, file storage backend, `write_through`, `wait_complete`, `hicache_size=80`

I searched open and closed issues/PRs for:

- `HiCache update_weights`
- `HiCache weight_version`
- `persistent storage weight update`
- `stale KV update_weights_from_tensor`
- `flush_cache_after_weight_update storage`
- `update_weights_from_disk HiCache`

I did not find an existing matching report.

**Comment by SouthWest7** [comment]
I'd like to help with.

**Comment by glaziermag** [comment]
@SouthWest7 Go for it!

**Comment by AlanFokCo** [comment]
We hit the same correctness issue in our cross-architecture model switching work. A few code-level observations that might help the fix:

**Root cause path:** `weight_updater.flush_cache()` → `scheduler.flush_cache()` → resets `tree_cache` and pool stats. But `tree_cache.reset()` only clears the in-memory radix tree; it doesn't touch the `HiCacheStorageBackend` that manages the persistent file layer. The persistent files survive, and the read path in `HiCacheStorageBackend.read()` has no weight-version check — it matches by token/page hash + model-name suffix, which stays the same after a weight update.

**Fix options (from simplest to most robust):**

1. **Minimal:** In `flush_cache()`, call `tree_cache.storage_backend.clear()` (or equivalent) to wipe persistent files. Cheap to implement, but loses all cached pages even if only one layer's weights changed.

2. **Version-keyed:** Include a `weight_version` (or hash of weight checksums) in the HiCache file key suffix. Currently the suffix is `_{model_name}_{tp_rank}_{pp_rank}` — appending `_{weight_version}` would make stale pages unmatchable. Old files can be lazily GC'd. This also handles the partial-update case (`update_weights_from_tensor` with a single layer) correctly — even a one-layer change invalidates all KV pages that depend on it.

3. **Layer-aware:** Track which layers changed and only invalidate pages whose prefix includes those layers. Most surgical but complex; probably not worth it for a first fix.

For reference, we went with option 1 in our setup (full invalidation on any weight change). The cost of cache rebuild is low compared to the correctness risk of serving with stale KV.

One edge case to watch: if `update_weights_from_distributed` (the NCCL path) also has `flush_cache=True`, the same fix needs to apply there. The `update_weights_from_tensor` path that @glaziermag reproduced with is just one of four weight-update entry points.
